toolchain/semantics/semantics_context.h

```diff
@@ -38,10 +38,15 @@ class SemanticsContext {
   // result.
   auto AddNodeAndPush(ParseTree::Node parse_node, SemanticsNode node) -> void;
 
-  // Adds a name to name lookup.
+  // Adds a name to name lookup. Prints a diagnostic for name conflicts.
   auto AddNameToLookup(ParseTree::Node name_node, SemanticsStringId name_id,
                        SemanticsNodeId target_id) -> void;
 
+  // Adds a name to name lookup. Ignores any name conflicts, with the
+  // expectation that they will have previously been diagnosed.
+  auto AddNameToLookupIgnoreConflicts(SemanticsStringId name_id,
+                                      SemanticsNodeId target_id) -> void;
```

Comment:
Is the expectation that there won't *be* conflicts because they've been previously diagnosed? If so, maybe assert?

If you imagine different callers needing to handle different cases, maybe it should return `bool` and let the caller handle it? That actually seems a bit better, and would allow you to call this from `AddNameToLookup` to share the code for inserting the name.

Reply:
> Is the expectation that there won't _be_ conflicts because they've been previously diagnosed? If so, maybe assert?

The expectation is only that they've been diagnosed. e.g., consider:

```
fn Foo(a: i32, a: i32) {
  return a + b;
}
```

The processing of the signature (`fn Foo(a: i32, a:i32)`) results in a diagnostics. When processing the definition, the names need to be added to name lookup. The 2nd argument is still present in IR even though it's a name conflict, so it gets re-added. The result is that we can produce diagnostics for the body, with `a` succeeding but `b` being a name lookup error.

For reference, the arguments can't just be "retained" because of things like `fn Foo(a: i32) -> a` which should fail name lookup. I would expect a deduced parameter to be valid in the return type (which it might not be right now, but should be easy to fix).

> If you imagine different callers needing to handle different cases, maybe it should return `bool` and let the caller handle it? That actually seems a bit better, and would allow you to call this from `AddNameToLookup` to share the code for inserting the name.

Will do, although `AddNameToLookup` is the only caller who should use it this way, per above.